Migrate controller implementation to kubeflow/common fashion

ChanYiLin opened the pull request:
Major changes include
1. Simplify v1 APIs by leverage common types in kubeflow/common/pkg/apis/common/v1
2. Replace exiting implementation with kubeflow/common/pkg/controller.v1/common
    - use `reconcileJob` from common library
    - use `reconcilePods` from common library
    - use `reconcileServices` from common library
    - replace `updateStatusSingle()` with new function `updateJobStatus()`
    - refactor TFController to make it consistent to controller interface in common library
3. Replace batch scheduler from kube-batch to volcano (this is alone with kubeflow common upgrade)
4. Remove `pkg/common/jobController`, `pkg/control`, `pkg/util`
5. refactor the test according to the change in the code base and make sure tf-operator pass almost all the old tests
    - remove `TestCleanupTFJob` since we cannot replace deleteTFJobHandler anymore with using common library

ChanYiLin commented:
/assign @Jeffwan @gaocegege 
Hi, please help me to review the PR, thanks!

ChanYiLin commented:
BTW, since I remove the `common/jobController`, `pkg/control` and `pkg/util` in tf-operator
If there is any other operator like pytorch or mxnet which is still using these libraries from tf-operator, they will need to change.

gaocegege commented:
/cc @johnugeorge @terrytangyuan

terrytangyuan commented:
> BTW, since I remove the `common/jobController`, `pkg/control` and `pkg/util` in tf-operator
> If there is any other operator like pytorch or mxnet which is still using these libraries from tf-operator, they will need to change.

If they are using older version of tf-operator code, it should be fine. They will be upgraded and switching to use kubeflow/common as well.

terrytangyuan commented on pkg/controller.v1/tensorflow/status_test.go:
```diff
@@ -442,60 +450,80 @@ func TestStatus(t *testing.T) {
+		if err != nil {
+			t.Errorf("%s: Expected error %v to be nil", c.description, err)
 		}
+		// if _, ok := c.tfJob.Spec.TFReplicaSpecs[tfv1.TFReplicaTypeChief]; ok {
```
Is this PR still in-progress? If not, delete the commented code.

terrytangyuan commented on pkg/controller.v1/tensorflow/status_test.go:
```diff
@@ -16,16 +16,22 @@
+	volcanoclient "volcano.sh/volcano/pkg/client/clientset/versioned"
+
 	common "github.com/kubeflow/common/pkg/apis/common/v1"
+	commonv1 "github.com/kubeflow/common/pkg/apis/common/v1"
```
Duplicate import here?

Jeffwan commented on hack/update-codegen.sh:
```diff
@@ -42,5 +42,5 @@ ${GOPATH}/bin/defaulter-gen  --input-dirs github.com/kubeflow/tf-operator/pkg/ap
 
 echo "Generating OpenAPI specification for tensorflow/v1"
-${GOPATH}/bin/openapi-gen --input-dirs github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1,k8s.io/api/core/v1,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/util/intstr,k8s.io/apimachinery/pkg/version --output-package github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1 --go-header-file hack/boilerplate/boiler ...
+${GOPATH}/bin/openapi-gen --report-filename=hack/violation_exception.list --input-dirs github.com/kubeflow/tf-operator/pkg/apis/tensorflow/v1,k8s.io/api/core/v1,k8s.io/apimachinery/pkg/apis/meta/v1,k8s.io/apimachinery/pkg/api/resource,k8s.io/apimachinery/pkg/runtime,k8s.io/apimachinery/pkg/util/intstr,k8s.io/apimachinery/pkg/version --output-package github.com/kubeflow/tf-operator/pkg/apis/tensor ...
```
em. Seems we assume `openapi-gen` is in the path. This will break for new users. I will create an issue to track to improve this later

ChanYiLin replied:
sure!

Jeffwan commented on hack/violation_exception.list:
```diff
@@ -0,0 +1,42 @@
+API rule violation: names_match,k8s.io/api/core/v1,AzureDiskVolumeSource,DataDiskURI
```
This is great.

Jeffwan commented on pkg/apis/tensorflow/v1/defaults_test.go:
```diff
@@ -55,10 +55,12 @@ func expectedTFJob(cleanPodPolicy common.CleanPodPolicy, restartPolicy common.Re
-			CleanPodPolicy: &cleanPodPolicy,
-			TFReplicaSpecs: map[TFReplicaType]*common.ReplicaSpec{
-				TFReplicaTypeWorker: &common.ReplicaSpec{
+			SuccessPolicy: &defaultSuccessPolicy,
```
I am not sure if success policy is part of run policy. This is something we can consolidate together in future releases

ChanYiLin replied:
Currently, SuccessPolicy and FailurePolicy(we have opened an issue for it) are for distributed tensorflow only, I am not sure the definition of Success and Failure are the same for other framework.

Jeffwan commented on pkg/common/util/v1/testutil/pod.go:
```diff
@@ -27,8 +27,8 @@ import (
 
 const (
 	// labels for pods and servers.
-	tfReplicaTypeLabel  = "tf-replica-type"
```
For test, I think this is fine. We should be cautious for public facing labels. I think some users may have logics depends on it.

ChanYiLin replied:
Actually, since we now use `reconcilePod` from common library, the labels are defined in common library
https://github.com/kubeflow/common/blob/master/pkg/apis/common/v1/constants.go#L5
and added to the pods here
https://github.com/kubeflow/common/blob/master/pkg/controller.v1/common/pod.go#L436
As you can see, it is hard coded now. 
That means, no matter in tf-operator or mxnet-operator, the labels in the pod will always be `apiv1.ReplicaIndexLabel = "replica-index"` and not the one defined in each operator.
I think we should modify the common library to use the `GetReplicaTypeLabelKey`
```
func (tc *TFController) GetReplicaTypeLabelKey() string {
	return tfReplicaTypeLabel
}
```

terrytangyuan replied:
I created separate issue to track this https://github.com/kubeflow/common/issues/98

Jeffwan replied:
I think we can have an interface there, and leave `replica-type` as default implementation in `kubeflow/common`, different operators can still override it if there's a need. 

This seems like a migration issue for tf only. Technically, we only use this label (and others) to filter pods with prefilter type. even they have same value, it's not a problem and won't affect filtering logic.

gaocegege replied:
Will it affect Katib? /cc @andreyvelich

andreyvelich replied:
Yes, in Katib we use labels to search for master pod.

@Jeffwan @terrytangyuan In the new `common` implementation, it still applies `job-role: master` for master pod ?

ChanYiLin replied:
Yes for job-role label, it is still `job-role:master`
But for replica type label, it has changed and removed the tf-  prefix
I will solve the issue in common library so we can overwrite the replica type label 
 and then change this back in tf-operator

andreyvelich replied:
Thanks @ChanYiLin! I believe it should work for us.

Jeffwan commented on pkg/control/pod_control.go:
```diff
@@ -1,177 +0,0 @@
-/*
```
This is great. We retire pod/service control package here.

Jeffwan commented on pkg/controller.v1/tensorflow/pod.go:
```diff
@@ -17,233 +17,49 @@ package tensorflow
 
-	gangSchedulingPodGroupAnnotation = "scheduling.k8s.io/group-name"
+	//gangSchedulerName = "volcano"
+	//gangSchedulingPodGroupAnnotation = "scheduling.k8s.io/group-name"
```
These variables are not being used now?

ChanYiLin replied:
yes, these are originally used in `reconcilePod` in tf-operator

Jeffwan commented on pkg/controller.v1/tensorflow/pod.go:
```diff
@@ -17,233 +17,49 @@ package tensorflow
+	//podTemplateRestartPolicyReason = "SettedPodTemplateRestartPolicy"
 	// exitedWithCodeReason is the normal reason when the pod is exited because of the exit code.
-	exitedWithCodeReason = "ExitedWithCode"
+	//exitedWithCodeReason = "ExitedWithCode"
```
It's safe to delete them. I think kubeflow/common reconcilePods() expose some these details. If not, please add to common.

Jeffwan commented on pkg/controller.v1/tensorflow/pod.go:
```diff
@@ -17,233 +17,49 @@ package tensorflow
-
-			// if master pod is present, select the master pod
-			// if master is not present, first worker pod is selected as the master.
-			if ContainChieforMasterSpec(tfjob) {
```
This is TF specific. Did you change to use `isMasterRole` https://github.com/kubeflow/common/blob/f77924701cb5d8a2d6b7f49f7afaec3c6271782a/pkg/apis/common/v1/interface.go#L57

ChanYiLin replied:
Yes, I have implemented it in controller.go L409

Jeffwan commented on pkg/controller.v1/tensorflow/pod.go:
```diff
@@ -292,19 +108,92 @@ func isDistributed(tfjob *tfv1.TFJob) bool {
-		if spec.Template.Spec.SchedulerName != "" && spec.Template.Spec.SchedulerName != tc.Config.GangSchedulerName {
-			return true
+// IsWorker0Completed return true if pod of worker0 succeeded and exited with 0
+func (tc *TFController) IsWorker0Completed(tfjob *tfv1.TFJob, replicas map[commonv1.ReplicaType]*commonv1.ReplicaSpec) (bool, error) {
```
I can not find reference for this methods. Can you point me to the caller?

ChanYiLin replied:
I use it and `PodRestart` in `UpdateJobStatus`.
Since in th-operator, we sometimes update status based on specific pod status (Chief, Master and worker0). With using the framework  from common library, we cannot use updateStatusSingle with flag `restart`, `worker0completed` to updateStatus.
In common library, we now updateStatus after reconcile all pods. So I have to implement these two function to iterate all pods again for specific  scenario.
(That is why I want to know how does mxnet-operator deal with scheduler succeeded in v1 with using common library)

Jeffwan commented on pkg/controller.v1/tensorflow/pod.go:
```diff
@@ -292,19 +108,92 @@ func isDistributed(tfjob *tfv1.TFJob) bool {
+		return false, err
+	}
+
+	podSlices := tc.GetPodSlices(pods, int(*spec.Replicas), logger)
```
minor: the logic here to get list of pods and extract pod status is very close to `isWorker0Complete`. Probably can do some refactor to reuse same code snippet.

Jeffwan commented:
@ChanYiLin Seems failure is from lint. Can you try to lint the code locally?

ChanYiLin commented:
Sure, but it is weird that the lint error cames from the file I didn't modify

terrytangyuan commented:
@ChanYiLin Maybe one of the dependencies' version got changed. Try and see if you can reproduce the issue locally.

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

gaocegege commented:
/test kubeflow-tf-operator-presubmit

gaocegege commented on build/images/tf_operator/Dockerfile:
```diff
@@ -4,6 +4,8 @@ ADD . /go/src/github.com/kubeflow/tf-operator
 
 WORKDIR /go/src/github.com/kubeflow/tf-operator
 
+RUN go mod download
```
I think we have the vendor in the git repo.

ChanYiLin replied:
yeap, but for common library, there is only `common/pkg/apis/common/v1` api definition in the vendor folder
but no `common/pkg/controller.v1/common` (which include `reconcileJob`, `reconcilePod`... these function)

ChanYiLin commented:
@gaocegege I found the reason of the test failed
we have no `glog` in the code base now
so we have to remove the arguments in the deployment yaml file ... including the test and the kubeflow/manifest

ChanYiLin commented:
hmm ... we have to handle the python sdk since we change the the TFJob api definition

ChanYiLin commented:
@jinchihe Hi!
We are now migrating the tf-operator to use v0.3.1 kubeflow/common
and I met an error from python e2e test, as you can see here
http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df?tab=workflow&nodeId=kubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df-2863222018&sidePanel=logs%3Akubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df-2863222018%3Amain

Then I found that current python/sdk still uses pretty old version of kubeflow/common
For example, [hack/python-sdk/gen-sdk.sh#L36](https://github.com/kubeflow/tf-operator/blob/master/hack/python-sdk/gen-sdk.sh#L36) is still using `github.com/kubeflow/common/job_controller/api/v1` which has already been removed in v0.3.1

Could you help me to update this part (python sdk and e2e test), so I can rebase and pass the test here?
Thanks!

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
> @jinchihe Hi!
> We are now migrating the tf-operator to use v0.3.1 kubeflow/common
> and I met an error from python e2e test, as you can see here
> http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df?tab=workflow&nodeId=kubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df-2863222018&sidePanel=logs%3Akubeflow-tf-operator-presubmit-v1-1171-58f4cb1-3696-f5df-2863222018%3Amain
> 
> Then I found that current python/sdk is still using pretty old version of kubeflow/common
> For example, [hack/python-sdk/gen-sdk.sh#L36](https://github.com/kubeflow/tf-operator/blob/master/hack/python-sdk/gen-sdk.sh#L36) is still using `github.com/kubeflow/common/job_controller/api/v1` which has already been removed in v0.3.1
> 
> Could you help me to update this part (python sdk and e2e test), so I can rebase and pass the test here?
> Thanks!

I found the reason why it failed, and fixed it here 
It doesn't require to update python sdk here

jinchihe commented:
> I found the reason why it failed, and fixed it here
> It doesn't require to update python sdk here

OK, please let me know if you need help, thanks.

johnugeorge commented:
@ChanYiLin  These are breaking changes in API which might have issues in other integrated repos. I feel, we should wait. 

/cc @gaocegege

ChanYiLin commented:
> @ChanYiLin These are breaking changes in API which might have issues in other integrated repos. I feel, we should take it to next release after 1.1
> 
> /cc @gaocegege

Agree, I don't think we have enough time to make it in release 1.1
There is still a lot of things to do in other repos which has dependencies with tf-operator

ChanYiLin commented:
> @ChanYiLin These are breaking changes in API which might have issues in other integrated repos. I feel, we should take it to next release after 1.1
> 
> /cc @gaocegege

Agree, I don't think we have enough time to make it in release 1.1
There is still a lot of things to do in other repos which has dependencies with tf-operator

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

gaocegege commented:
/retest

Jeffwan commented:
@ChanYiLin Can you generate the codes in your local env successfully?

Jeffwan commented:
Agree, I think we can postpone this change to next release. Kubeflow 1.1 will come with all stable operator versions.

ChanYiLin commented:
@Jeffwan It passed most of e2e tests which means the code was built and ran successfully, but several tests got timeout error.
It means something happened in the real cluster env, but I cant get the log of controller from our testing.
I will try to run it in my own k8s env to see the log.

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
taOk, I found the reason why it failed.
Still because we remove updateStatusSingle and do updateStatus after reconcile all replicas.
So the controller won't know which one has been restart (we pass restart as arg to updateStatusSingle  before)
and will consider it as failed...
I originally try to get all the pods again in updateStatus to see whether there is a pod need to be restarted
but this won't work because it has already be delete while reconcile replicas...
need to figure out another way

terrytangyuan commented:
@ChanYiLin Any update on this?

ChanYiLin commented:
> @ChanYiLin Any update on this?

Yes, I have tried for a while.
I think there is no other way but have to reimplement reconcilePod in tf-operator (rather than use the one from common lib),
so we can know whether there is a pod restarted and update to correct status.
I will try this today

ChanYiLin commented:
@terrytangyuan @gaocegege  @Jeffwan  The PR is basically ready to merge.
For testing, if we test simple-gpu-tfjob test case, it will get stucked until timeout and affect other test case (resource not enough), but if I remove the test, it will pass all other test cases. 
Thus, I think there is no functionality issue in this PR change now, it can pass all e2e tests we design before(clean pod policy, restart policy, etc).

Jeffwan commented:
@ChanYiLin This is more due to resource outage? I think if that's the problem. it's fine. reconciler logic is same for cpu and gpu workloads. I am fine moving forward without GPU tests. This can be done with manual testing instead if that's resource outage.

ChanYiLin commented:
Before, theres was an issue about restart policy, but I fixed it now.
Yes, as you can see, after removing the gpu test, it passed the last 30 e2e tests cases (kubeflow-tf-operator-presubmit).
I agree that we can merge it first and test gpu case manually. 
But before we merge this PR, I would like to add the gpu tests back and squash the commits.

gaocegege commented:
@jlewi Can we have more resources in the future?

terrytangyuan commented:
LGTM as well. We can discuss resource scarcity in a separate issue. 

/lgtm

ChanYiLin commented:
Hi, I've squashed the commit and added simple-tfjob-gpu e2e test back.
It might fail because of gpu test, but I think we can merge it first for it has passed last 30 e2e tests before.

gaocegege commented:
@ChanYiLin Can we disable the test if it cannot pass?

gaocegege commented on cmd/tf-operator.v1/app/server.go:
```diff
@@ -26,9 +26,9 @@ import (
 	controller "github.com/kubeflow/tf-operator/pkg/controller.v1/tensorflow"
-	"github.com/kubeflow/tf-operator/pkg/util/signals"
+
+	"github.com/kubeflow/common/pkg/util/signals"
```
Can we reorder the groups? We suggest placing third-party deps at the second group

ChanYiLin commented:
@gaocegege you mean comment out the test temporarily?

gaocegege commented on pkg/controller.v1/tensorflow/pod_test.go:
```diff
@@ -502,10 +543,11 @@ func TestScaleUp(t *testing.T) {
-	if err != nil {
-		t.Errorf("%s: unexpected error when syncing jobs %v", tfJob.Name, err)
-	}
+	_ = ctr.ReconcileJobs(tfJob, tfJob.Spec.TFReplicaSpecs, tfJob.Status, &tfJob.Spec.RunPolicy)
```
Why don't we check the error here?

gaocegege commented:
> you mean comment out the test temporarily?

Yeah. We cannot merge the PR if CI fails. We can open an issue to keep track of resource problems and disable it temporarily

ChanYiLin commented:
@gaocegege  it passed the presubmit test. 
For CI, I check hack/verify-codegen.sh locally and it passed, don't know why it got issue here.

gaocegege commented:
Maybe it is caused by the version beween your env and CI env. Can you try to upgrade your codegen?

ChanYiLin commented:
/retest

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/retest

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
@gaocegege @terrytangyuan do you have any idea about this issue
The e2e test keeps failing at "setup kubeflow" and here is the log
I think it is not from tf-operator
```
INFO:root:Running: ks-13 env add e2e-0729-1322-e4a4 cwd=/mnt/test-data-volume/kubeflow-tf-operator-presubmit-v1-1171-232e11d-8096-3703/src/kubeflow/tf-operator/test/test-app 
INFO:root:Subprocess output: 
INFO:root:level=info msg="Using context \"gke_kubeflow-ci_us-east1-d_ztor-presubmit-v1-1171-232e11d-8096-3703\" from kube config file \"/mnt/test-data-volume/kubeflow-tf-operator-presubmit-v1-1171-232e11d-8096-3703/.kube/kubeconfig\"" INFO:root:level=info msg="Creating environment \"e2e-0729-1322-e4a4\" with namespace \"default\", pointing to \"version:v 1.15.12\" cluster at address \"https://35.196.25.75\"" 
INFO:root:level=error msg="No Major.Minor.Patch elements found"
```

ChanYiLin commented:
/retest

gaocegege commented:
/retest

Jeffwan commented:
Interesting. 

http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-tf-operator-presubmit-v1-1171-232e11d-8256-de7d?tab=workflow

I didn't see any test component has failures. However, the parent component is red.

In the logs

```
ERROR|2020-07-30T12:48:16|/src/kubeflow/testing/py/kubeflow/testing/run_e2e_workflow.py|483| Exception occurred: Timeout waiting for workflows kubeflow-tf-operator-presubmit-v1-1171-232e11d-8256-de7d in namespace kubeflow-test-infra to finish
Traceback (most recent call last):
  File "/src/kubeflow/testing/py/kubeflow/testing/run_e2e_workflow.py", line 474, in run
    status_callback=argo_client.log_status
  File "/src/kubeflow/testing/py/kubeflow/testing/argo_client.py", line 157, in wait_for_workflows
    raise util.ExceptionWithWorkflowResults(message, all_results)
ExceptionWithWorkflowResults: Timeout waiting for workflows kubeflow-tf-operator-presubmit-v1-1171-232e11d-8256-de7d in namespace kubeflow-test-infra to finish

```

The job takes ~1hr and didn't go into ending state. We probably need to run this image locally and see why job status doesn't change

jinchihe commented:
I think that's caused by no space left on the test-data-volume. @Jeffwan 
```
INFO:root:level=error msg="mkdir /mnt/test-data-volume/kubeflow-tf-operator-presubmit-v1-1179-c81bb9a-0720-04a8/src/kubeflow/tf-operator/
test/test-app/environments/e2e-0805-0709-3019: no space left on device"
```

jinchihe commented:
Waiting for a moment, I'm going to handle this in https://github.com/kubeflow/testing/issues/742

jinchihe commented:
/retest

jinchihe commented:
Another strange error while setup cluster in CI tests.
```
INFO:root:Using ksonnet cmd: ks-13
INFO:root:Running: ks-13 env add e2e-0806-0218-ca83
cwd=/mnt/test-data-volume/kubeflow-tf-operator-presubmit-v1-1171-232e11d-7216-b6d9/src/kubeflow/tf-operator/test/test-app
INFO:root:Subprocess output:
INFO:root:level=info msg="Using context \"gke_kubeflow-ci_us-east1-d_ztor-presubmit-v1-1171-232e11d-7216-b6d9\" from kubeconfig file \"/m
nt/test-data-volume/kubeflow-tf-operator-presubmit-v1-1171-232e11d-7216-b6d9/.kube/kubeconfig\""
INFO:root:level=info msg="Creating environment \"e2e-0806-0218-ca83\" with namespace \"default\", pointing to \"version:v1.15.12\" cluste
r at address \"https://35.185.20.182\""
INFO:root:level=error msg="No Major.Minor.Patch elements found"
```

jinchihe commented:
The `No Major.Minor.Patch elements found` problem already fixed in https://github.com/kubeflow/tf-operator/pull/1179. 
@ChanYiLin  You can rebase once the PR https://github.com/kubeflow/tf-operator/pull/1179 merged. Thanks.

ChanYiLin commented:
/retest

ChanYiLin commented:
@jinchihe there is still another issue in setup-kubeflow and  the ci also failed. Can you help to verify the issue? Thanks

ChanYiLin commented:
/retest

ChanYiLin commented:
/retest

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/retest

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
/test kubeflow-tf-operator-presubmit

ChanYiLin commented:
@gaocegege  @Jeffwan This PR finally passed all the tests :)
Could you help me to approve it. Thanks!

ChanYiLin commented:
let me squash the commit.

gaocegege commented:
Thanks for your contribution! :tada: :+1:

gaocegege commented:
/lgtm

Jeffwan commented:
@ChanYiLin I don't have permission in this repo. I think you can self approve it now. Great work!

Jeffwan commented:
-> "cla/google Expected — Waiting for status to be reported"

There is some problem with the network traffic and CLA Assistant took some time to respond.

I manually call`GET https://cla-assistant.io/check/kubeflow/tf-operator?pullrequest=1171` retrigger the CLA Assistant check If CLA Assistant takes time to respond.

Jeffwan commented:
/test kubeflow-tf-operator-presubmit

gaocegege commented:
/retest
